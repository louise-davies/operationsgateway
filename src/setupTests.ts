// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import crypto from 'crypto';
// TODO JOEL: Is this still needed? - Also in vite config
// need to mock <canvas> for plotting
import 'vitest-canvas-mock';
// TODO JOEL: Replace with vitest-fail-on-console
import failOnConsole from 'jest-fail-on-console';
import { TextEncoder } from 'util';
import { server } from './mocks/server';

// TODO JOEL: Is this still needed? - same for failOnConsole below
global.TextEncoder = TextEncoder;

failOnConsole();

vi.setConfig({ testTimeout: 15000 });

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());

// TODO JOEL: Check if a all of the below is still needed

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {},
  });
}

// this is needed because of https://github.com/facebook/jest/issues/8987
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let mockActualReact;
vi.doMock('react', async () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (!mockActualReact) {
    mockActualReact = await vi.importActual('react');
  }
  return mockActualReact;
});

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.revokeObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {
      // no-op
    },
  });
}

// jest doesn't implement web crypto so set up nodejs crypto as a default
Object.defineProperty(global, 'crypto', {
  value: Object.setPrototypeOf({ subtle: crypto.subtle }, crypto),
});
