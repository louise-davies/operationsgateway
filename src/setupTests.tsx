// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
// need to mock <canvas> for plotting
import 'jest-canvas-mock';
import {
  FullChannelMetadata,
  PlotDataset,
  timeChannelName,
  FullScalarChannelMetadata,
  DEFAULT_WINDOW_VARS,
} from './app.types';
import { Action, PreloadedState, ThunkAction } from '@reduxjs/toolkit';
import { AppStore, RootState, setupStore } from './state/store';
import { initialState as initialConfigState } from './state/slices/configSlice';
import { initialState as initialTableState } from './state/slices/tableSlice';
import { initialState as initialSearchState } from './state/slices/searchSlice';
import {
  initialState as initialPlotState,
  PlotConfig,
} from './state/slices/plotSlice';
import { initialState as initialFilterState } from './state/slices/filterSlice';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { QueryClientProvider, QueryClient } from '@tanstack/react-query';
import { COLOUR_ORDER } from './plotting/plotSettings/colourGenerator';
import { staticChannels } from './api/channels';
import { server } from './mocks/server';
import { matchRequestUrl, MockedRequest } from 'msw';
import channelsJson from './mocks/channels.json';

jest.setTimeout(15000);

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());

/**
 * Waits for msw request -
 * @param method string representing the HTTP method
 * @param url string representing the URL match for the route
 * @returns a promise of the matching request
 *  */
export function waitForRequest(method: string, url: string) {
  let requestId = '';

  return new Promise<MockedRequest>((resolve, reject) => {
    server.events.on('request:start', (req) => {
      const matchesMethod = req.method.toLowerCase() === method.toLowerCase();

      const matchesUrl = matchRequestUrl(req.url, url).matches;

      if (matchesMethod && matchesUrl) {
        requestId = req.id;
      }
    });

    server.events.on('request:match', (req) => {
      if (req.id === requestId) {
        resolve(req);
      }
    });

    server.events.on('request:unhandled', (req) => {
      if (req.id === requestId) {
        reject(
          new Error(`The ${req.method} ${req.url.href} request was unhandled.`)
        );
      }
    });
  });
}

// this is needed because of https://github.com/facebook/jest/issues/8987
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let mockActualReact;
jest.doMock('react', () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (!mockActualReact) {
    mockActualReact = jest.requireActual('react');
  }
  return mockActualReact;
});

export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getInitialState = (): RootState => ({
  config: initialConfigState,
  table: initialTableState,
  search: initialSearchState,
  plots: initialPlotState,
  filter: initialFilterState,
});
export const dispatch = (
  action: Action | ThunkAction<void, RootState, unknown, Action<string>>
): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getInitialState, null);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.revokeObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {
      // no-op
    },
  });
}

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
  queryClient?: QueryClient;
}

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
    logger: {
      log: console.log,
      warn: console.warn,
      error: jest.fn(),
    },
  });

export const hooksWrapperWithProviders = (
  state = {},
  queryClient?: QueryClient
) => {
  const testQueryClient = queryClient ?? createTestQueryClient();
  const store = setupStore(state);
  const wrapper = ({ children }) => (
    <Provider store={store}>
      <QueryClientProvider client={testQueryClient}>
        {children}
      </QueryClientProvider>
    </Provider>
  );
  return wrapper;
};

export function renderComponentWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </Provider>
    );
  }
  return {
    store,
    queryClient,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export function renderComponentWithStore(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => undefined,
      addEventListener: () => undefined,
      removeEventListener: () => undefined,
      addListener: () => undefined,
      removeListener: () => undefined,
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};

export const testChannels = [
  ...Object.values(staticChannels),
  ...Object.entries(channelsJson.channels).map(
    ([systemName, channel]) =>
      ({
        systemName,
        ...channel,
      } as FullChannelMetadata)
  ),
];

export const testScalarChannels: FullScalarChannelMetadata[] = [
  ...Object.values(staticChannels),
  ...Object.entries(channelsJson.channels)
    .filter(([systemName, channel]) => channel.type === 'scalar')
    .map(
      ([systemName, channel]) =>
        ({
          systemName,
          ...channel,
        } as FullScalarChannelMetadata)
    ),
];

export const generatePlotDataset = (num: number) => {
  const datasetName = testScalarChannels[num].systemName;
  const plotDataset: PlotDataset = {
    name: datasetName,
    data: [
      {
        timestamp: num,
        [datasetName]: num + num,
      },
      {
        timestamp: num + num,
        [datasetName]: num + num + num,
      },
      {
        timestamp: num + num + num,
        [datasetName]: num + num + num + num,
      },
    ],
  };
  return plotDataset;
};

export const testPlotDatasets = Array.from(Array(3), (_, i) =>
  generatePlotDataset(i + 1)
);

export const generatePlotConfig = (num: number) => {
  const plotTitle = `Plot ${num}`;

  const plotConfig: PlotConfig = {
    open: num % 2 === 0,
    title: plotTitle,
    plotType: num % 2 === 0 ? 'scatter' : 'line',
    XAxis: num % 3 === 0 ? timeChannelName : undefined,
    XAxisScale:
      num % 3 === 0 ? 'time' : num % 3 === 1 ? 'linear' : 'logarithmic',
    selectedPlotChannels: [],
    leftYAxisScale: num % 2 === 0 ? 'linear' : 'logarithmic',
    rightYAxisScale: num % 2 === 0 ? 'logarithmic' : 'linear',
    gridVisible: num % 2 === 0,
    axesLabelsVisible: num % 2 !== 0,
    selectedColours: [],
    remainingColours: COLOUR_ORDER.map((colour) => colour),
    ...DEFAULT_WINDOW_VARS,
  };

  return plotConfig;
};

export const testPlotConfigs = Array.from(Array(3), (_, i) =>
  generatePlotConfig(i)
);
