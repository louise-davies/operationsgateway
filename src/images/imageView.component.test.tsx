import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import ImageView, { ImageViewProps } from './imageView.component';
import { flushPromises } from '../setupTests';

describe('Image view component', () => {
  let props: ImageViewProps;

  beforeAll(() => {
    // mock to ensure img.onload gets called to set the canvas width & height
    // modified from: https://github.com/jsdom/jsdom/issues/1816#issuecomment-432496573
    Object.defineProperty(global.Image.prototype, 'src', {
      set(src) {
        setTimeout(() => this.onload());
      },
    });
    Object.defineProperty(global.Image.prototype, 'height', {
      get: () => 200,
    });
    Object.defineProperty(global.Image.prototype, 'width', {
      get: () => 300,
    });
  });

  beforeEach(() => {
    props = {
      image: 'testSrc',
      title: 'Test image',
      viewReset: false,
    };
  });

  it('renders an image and a canvas overlay', async () => {
    const { asFragment } = render(<ImageView {...props} />);

    // "load" image
    await flushPromises();

    expect(asFragment()).toMatchSnapshot();
  });

  it('can zoom into image', async () => {
    render(<ImageView {...props} />);

    // "load" image
    await flushPromises();

    const overlay = screen.getByTestId('overlay');
    const image = screen.getByAltText('Test image');

    fireEvent.mouseDown(overlay, { button: 0, clientX: 0, clientY: 0 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 100, clientY: 100 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 100, clientY: 100 });

    expect(image).toHaveStyle({
      transform: 'translate(0px,0px) scale(2,2)',
    });
    fireEvent.mouseDown(overlay, { button: 0, clientX: 100, clientY: 100 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 40, clientY: 80 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 0, clientY: 0 });

    expect(image).toHaveStyle({
      transform: 'translate(-200px,-300px) scale(5,5)',
    });
  });

  it('can pan around zoomed image', async () => {
    render(<ImageView {...props} />);

    // "load" image
    await flushPromises();

    const overlay = screen.getByTestId('overlay');
    const image = screen.getByAltText('Test image');

    // zoom into image
    fireEvent.mouseDown(overlay, { button: 0, clientX: 100, clientY: 100 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 0, clientY: 0 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 0, clientY: 0 });

    expect(image).toHaveStyle({
      transform: 'translate(0px,0px) scale(2,2)',
    });

    fireEvent.mouseDown(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 100,
      clientY: 100,
    });
    fireEvent.mouseMove(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 50,
      clientY: 50,
    });
    fireEvent.mouseUp(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 50,
      clientY: 50,
    });

    expect(image).toHaveStyle({
      transform: 'translate(-50px,-50px) scale(2,2)',
    });
  });

  it('resets view when viewReset is changed', async () => {
    const { rerender } = render(<ImageView {...props} />);

    // "load" image
    await flushPromises();

    const overlay = screen.getByTestId('overlay');
    const image = screen.getByAltText('Test image');

    fireEvent.mouseDown(overlay, { button: 0, clientX: 100, clientY: 100 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 50, clientY: 100 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 50, clientY: 100 });

    expect(image).toHaveStyle({
      transform: 'translate(-300px,-600px) scale(6,6)',
    });

    props.viewReset = !props.viewReset;
    rerender(<ImageView {...props} />);

    expect(image).toHaveStyle({
      transform: 'translate(0px,0px) scale(1,1)',
    });
  });

  it('ignores non-left mouse clicks and mouse move events without corresponding mouse down events', async () => {
    render(<ImageView {...props} />);

    // "load" image
    await flushPromises();

    const overlay = screen.getByTestId('overlay');
    const image = screen.getByAltText('Test image');

    fireEvent.mouseDown(overlay, { button: 2, clientX: 50, clientY: 100 });

    expect(image).toHaveStyle({
      transform: 'translate(0px,0px) scale(1,1)',
    });

    fireEvent.mouseMove(overlay, { button: 0, clientX: 50, clientY: 100 });

    expect(image).toHaveStyle({
      transform: 'translate(0px,0px) scale(1,1)',
    });
  });

  it("doesn't let you pan out of bounds", async () => {
    render(<ImageView {...props} />);

    // "load" image
    await flushPromises();

    const overlay = screen.getByTestId('overlay');
    const image = screen.getByAltText('Test image');

    fireEvent.mouseDown(overlay, { button: 0, clientX: 75, clientY: 50 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 225, clientY: 150 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 225, clientY: 150 });

    expect(image).toHaveStyle({
      transform: 'translate(-150px,-100px) scale(2,2)',
    });

    // try to pan out of bounds past 300,200
    fireEvent.mouseDown(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 200,
      clientY: 200,
    });
    fireEvent.mouseMove(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 0,
      clientY: 0,
    });
    fireEvent.mouseUp(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 0,
      clientY: 0,
    });

    expect(image).toHaveStyle({
      transform: 'translate(-150px,-100px) scale(2,2)',
    });

    // try to pan out of bounds past 0,0
    fireEvent.mouseDown(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 0,
      clientY: 0,
    });
    fireEvent.mouseMove(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 200,
      clientY: 200,
    });
    fireEvent.mouseUp(overlay, {
      shiftKey: true,
      button: 0,
      clientX: 200,
      clientY: 200,
    });

    expect(image).toHaveStyle({
      transform: 'translate(0px,0px) scale(2,2)',
    });
  });

  it("doesn't let you zoom out of bounds", async () => {
    const { rerender } = render(<ImageView {...props} />);

    // "load" image
    await flushPromises();

    const overlay = screen.getByTestId('overlay');
    const image = screen.getByAltText('Test image');

    fireEvent.mouseDown(overlay, { button: 0, clientX: 75, clientY: 50 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 225, clientY: 150 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 225, clientY: 150 });

    expect(image).toHaveStyle({
      transform: 'translate(-150px,-100px) scale(2,2)',
    });

    // try to zoom out of bounds past 300,200
    fireEvent.mouseDown(overlay, { button: 0, clientX: 290, clientY: 190 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 300, clientY: 200 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 300, clientY: 200 });

    expect(image).toHaveStyle({
      transform: 'translate(-5700px,-3800px) scale(20,20)',
    });

    props.viewReset = !props.viewReset;
    rerender(<ImageView {...props} />);

    // try to zoom out of bounds past 0 on x axis
    fireEvent.mouseDown(overlay, { button: 0, clientX: 2, clientY: 100 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 1, clientY: 0 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 1, clientY: 0 });

    expect(image).toHaveStyle({
      transform: 'translate(0px,0px) scale(2,2)',
    });

    props.viewReset = !props.viewReset;
    rerender(<ImageView {...props} />);

    // try to zoom out of bounds past 0 on Y axis
    fireEvent.mouseDown(overlay, { button: 0, clientX: 100, clientY: 2 });
    fireEvent.mouseMove(overlay, { button: 0, clientX: 0, clientY: 1 });
    fireEvent.mouseUp(overlay, { button: 0, clientX: 0, clientY: 1 });

    expect(image).toHaveStyle({
      transform: 'translate(0px,0px) scale(3,3)',
    });
  });
});
