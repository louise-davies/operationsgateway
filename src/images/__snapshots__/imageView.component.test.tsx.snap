// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image view component > renders an image and a canvas overlay 1`] = `
<DocumentFragment>
  <div
    style="position: relative;"
  >
    <canvas
      data-testid="overlay"
      height="200"
      style="position: absolute; z-index: 2; pointer-events: none;"
      width="300"
    />
    <div
      style="display: inline-block; overflow: hidden;"
    >
      <img
        alt="Test image"
        src="testSrc"
        style="transform: translate(0px,0px) scale(1); transform-origin: top left; image-rendering: pixelated;"
      />
    </div>
  </div>
</DocumentFragment>
`;
